File: packages/contract/src/playback.ts
/** playback start/seek/audio-track + heartbeat/continue-watching. */

import { z } from "zod";

export const PlaybackMethod = z.enum(["DIRECT_PLAY", "DIRECT_STREAM", "REMUX", "TRANSCODE"]);

export const DeviceProfile = z.object({
  supportedContainers: z.array(z.string()),
  supportedVideoCodecs: z.array(z.string()),
  supportedAudioCodecs: z.array(z.string()),
  maxVideoBitrateKbps: z.number().optional(),
  maxWidth: z.number().optional(),
  maxHeight: z.number().optional(),
  supportsHdr: z.boolean().optional(),
  subtitleMode: z.enum(["none", "external", "burn"]),
  enableDirectPlay: z.boolean().optional(),
  enableDirectStream: z.boolean().optional(),
});
export type DeviceProfile = z.infer<typeof DeviceProfile>;

export const StartPlaybackBody = z.object({
  profileId: z.string(),
  mediaItemId: z.string(),
  mediaFileId: z.string(),
  deviceProfile: DeviceProfile,
  subtitleTrackId: z.string().optional(),
  audioStreamIndex: z.number().optional(),
});
export type StartPlaybackBody = z.infer<typeof StartPlaybackBody>;

export const StartPlaybackResponse = z.object({
  sessionId: z.string(),
  method: PlaybackMethod,
  reasons: z.array(z.string()),
  playlistUrl: z.string().nullable(),
  /** REMUX only: the progressive fragmented-MP4 stream URL (native <video>, range requests). */
  streamUrl: z.string().nullable().default(null),
  /** Where playback should resume (ms) — 0 when starting fresh. Server-side state, so callers need no stored position. */
  resumePositionMs: z.number().default(0),
  /**
   * Media-absolute duration of the file (ms) — the player's playbar total.
   * Streams started mid-file (REMUX/TRANSCODE resume) report a shorter
   * element duration, so clients use this for the displayed end time while
   * seeking in stream-relative coordinates.
   */
  absoluteDurationMs: z.number().default(0),
  /**
   * Exact media time the stream starts at — the client's timeline offset.
   * TRANSCODE (accurate seek) starts at the raw resume/target position,
   * frame-exact; REMUX starts at the keyframe at-or-before it. The client
   * self-seeks from here to resumePositionMs so continue-watching lands on
   * the exact stored position.
   */
  actualStartMs: z.number().optional(),
});
export type StartPlaybackResponse = z.infer<typeof StartPlaybackResponse>;

export const PlaybackSessionParams = z.object({ sessionId: z.string() });

export const SeekBody = z.object({ positionMs: z.number() });
export type SeekBody = z.infer<typeof SeekBody>;

export const SeekResponse = z.object({
  restarted: z.boolean(),
  segmentFrom: z.number(),
  pid: z.number(),
  killedPid: z.number().optional(),
  /**
   * TRANSCODE (accurate seek) restarts land on the exact target; REMUX lands
   * on the keyframe at-or-before it — the position the restarted stream
   * actually starts at. The client's timeline offset uses this so subs and
   * positions stay exact after a restart.
   */
  actualStartMs: z.number().optional(),
});

export const AudioTrackSwitchBody = z.object({ audioStreamIndex: z.number(), positionMs: z.number() });
export type AudioTrackSwitchBody = z.infer<typeof AudioTrackSwitchBody>;

/** Quality switch — new encode caps. Omitted fields keep the session's current caps. */
export const QualitySwitchBody = z.object({
  /** Media-absolute position to restart from — the client converts its timeline time like /seek. */
  positionMs: z.number(),
  /** Drop forced caps and re-decide with the session's start profile — a capped
   *  TRANSCODE falls back up the ladder to DIRECT_PLAY when the source allows. */
  reset: z.boolean().optional(),
  maxWidth: z.number().optional(),
  maxHeight: z.number().optional(),
  maxVideoBitrateKbps: z.number().optional(),
});
export type QualitySwitchBody = z.infer<typeof QualitySwitchBody>;

/**
 * Quality-switch response — mirrors SeekResponse but carries the stream URLs
 * because the method itself can change (e.g. REMUX at 1080p -> TRANSCODE at
 * 480p, or a capped TRANSCODE -> DIRECT_PLAY on reset), and the client must
 * swap its src (streamUrl <-> playlistUrl <-> direct file).
 */
export const QualitySwitchResponse = z.object({
  restarted: z.boolean(),
  method: PlaybackMethod,
  segmentFrom: z.number().nullable(),
  pid: z.number().nullable(),
  killedPid: z.number().optional(),
  /**
   * TRANSCODE (accurate seek) restarts land on the exact target; REMUX lands
   * on the keyframe at-or-before it — the position the restarted stream
   * actually starts at.
   */
  actualStartMs: z.number().optional(),
  playlistUrl: z.string().nullable(),
  streamUrl: z.string().nullable(),
});

export const AudioTrackSwitchResponse = z.object({
  restarted: z.boolean(),
  segmentFrom: z.number(),
  pid: z.number(),
  killedPid: z.number(),
  /**
   * TRANSCODE (accurate seek) restarts land on the exact target; REMUX lands
   * on the keyframe at-or-before it — the position the restarted stream
   * actually starts at.
   */
  actualStartMs: z.number().optional(),
});

export const HeartbeatBody = z.object({
  positionMs: z.number(),
  durationMs: z.number().optional(),
});
export type HeartbeatBody = z.infer<typeof HeartbeatBody>;

export const HeartbeatResponse = z.object({ ok: z.boolean(), watched: z.boolean() });

export const StopResponse = z.object({ ok: z.boolean() });

export const ContinueWatchingQuery = z.object({ profileId: z.string() });
export type ContinueWatchingQuery = z.infer<typeof ContinueWatchingQuery>;

export const MediaItemRef = z.object({
  id: z.string(),
  kind: z.enum(["MOVIE", "SERIES", "SEASON", "EPISODE"]),
  title: z.string(),
  sortTitle: z.string(),
  parentId: z.string().nullable(),
  seasonNumber: z.number().nullable(),
  episodeNumber: z.number().nullable(),
  year: z.number().int().nullable(),
  posterUrl: z.string().nullable(),
  backdropUrl: z.string().nullable(),
  mediaFileId: z.string().nullable(),
});

export const ContinueWatchingEntry = z.object({
  mediaItem: MediaItemRef.passthrough(),
  /** Show title for an EPISODE mediaItem (the series it belongs to), null otherwise. */
  seriesTitle: z.string().nullable(),
  /** Detail page to land on — the series for an episode, the item itself otherwise. */
  detailItemId: z.string(),
  positionMs: z.number(),
  durationMs: z.number().nullable(),
  upNext: z.boolean(),
});
export type ContinueWatchingEntry = z.infer<typeof ContinueWatchingEntry>;
export const ContinueWatchingResponse = z.array(ContinueWatchingEntry);

/** One row per (profile, item, calendar day) — the day-by-day watch history. */
export const WatchHistoryQuery = z.object({ profileId: z.string(), mediaItemId: z.string() });
export type WatchHistoryQuery = z.infer<typeof WatchHistoryQuery>;

export const WatchHistoryEntry = z.object({
  date: z.coerce.date(),
  /** Total watch time credited that day (seek jumps excluded). */
  watchedMs: z.number(),
  /** Start of the day's first watch span. */
  firstStartedAt: z.coerce.date().nullable(),
  /** End of the day's last watch span. */
  lastEndedAt: z.coerce.date().nullable(),
  /** Completions (rewatch events) that day. */
  completions: z.number(),
});
export type WatchHistoryEntry = z.infer<typeof WatchHistoryEntry>;

export const WatchHistoryResponse = z.array(WatchHistoryEntry);

/** Manual watched-marking (right-click menu): per-profile, idempotent. */
export const SetWatchedParams = z.object({ mediaItemId: z.string() });
export type SetWatchedParams = z.infer<typeof SetWatchedParams>;

export const SetWatchedBody = z.object({
  profileId: z.string(),
  /** true = mark watched, false = reset to unwatched (position and count cleared). */
  watched: z.boolean().default(true),
});
export type SetWatchedBody = z.infer<typeof SetWatchedBody>;

export const SetWatchedResponse = z.object({ ok: z.boolean(), watched: z.boolean() });

export const ErrorResponse = z.object({ error: z.string() });

File: native/flutter/lib/core/api/models/playback.dart
/// Mirrors packages/contract/src/playback.ts.
class DeviceProfile {
  final List<String> supportedContainers;
  final List<String> supportedVideoCodecs;
  final List<String> supportedAudioCodecs;
  final int? maxVideoBitrateKbps;
  final int? maxWidth;
  final int? maxHeight;
  final bool? supportsHdr;
  final String subtitleMode; // none | external | burn
  final bool? enableDirectPlay;
  final bool? enableDirectStream;

  const DeviceProfile({
    required this.supportedContainers,
    required this.supportedVideoCodecs,
    required this.supportedAudioCodecs,
    this.maxVideoBitrateKbps,
    this.maxWidth,
    this.maxHeight,
    this.supportsHdr,
    required this.subtitleMode,
    this.enableDirectPlay,
    this.enableDirectStream,
  });

  Map<String, dynamic> toJson() => {
        'supportedContainers': supportedContainers,
        'supportedVideoCodecs': supportedVideoCodecs,
        'supportedAudioCodecs': supportedAudioCodecs,
        if (maxVideoBitrateKbps != null) 'maxVideoBitrateKbps': maxVideoBitrateKbps,
        if (maxWidth != null) 'maxWidth': maxWidth,
        if (maxHeight != null) 'maxHeight': maxHeight,
        if (supportsHdr != null) 'supportsHdr': supportsHdr,
        'subtitleMode': subtitleMode,
        if (enableDirectPlay != null) 'enableDirectPlay': enableDirectPlay,
        if (enableDirectStream != null) 'enableDirectStream': enableDirectStream,
      };

  /// media_kit (libmpv) plays effectively anything and renders ASS natively —
  /// the device profile exists so the server can still choose DIRECT_PLAY
  /// whenever possible (cheapest for the server) with wide codec/container
  /// support declared. burn-in subtitleMode: libmpv handles text/ASS subs
  /// itself once tracks are added, so we never need server burn-in for text
  /// formats — only bitmap (PGS/VOBSUB) forces it, same as the web client.
  static const native = DeviceProfile(
    supportedContainers: ['mp4', 'mkv', 'webm', 'mov'],
    supportedVideoCodecs: ['h264', 'hevc', 'vp9', 'av1'],
    supportedAudioCodecs: ['aac', 'ac3', 'eac3', 'opus', 'flac', 'mp3', 'dts'],
    supportsHdr: true,
    subtitleMode: 'external',
    enableDirectPlay: true,
    enableDirectStream: true,
  );
}

class StartPlaybackResponse {
  final String sessionId;
  final String method; // DIRECT_PLAY | DIRECT_STREAM | REMUX | TRANSCODE
  final String? playlistUrl;
  final String? streamUrl;
  final int resumePositionMs;
  final int absoluteDurationMs;
  final int? actualStartMs;

  StartPlaybackResponse({
    required this.sessionId,
    required this.method,
    required this.playlistUrl,
    required this.streamUrl,
    required this.resumePositionMs,
    required this.absoluteDurationMs,
    required this.actualStartMs,
  });

  factory StartPlaybackResponse.fromJson(Map<String, dynamic> j) => StartPlaybackResponse(
        sessionId: j['sessionId'] as String,
        method: j['method'] as String,
        playlistUrl: j['playlistUrl'] as String?,
        streamUrl: j['streamUrl'] as String?,
        resumePositionMs: j['resumePositionMs'] as int? ?? 0,
        absoluteDurationMs: j['absoluteDurationMs'] as int? ?? 0,
        actualStartMs: j['actualStartMs'] as int?,
      );
}

class SeekResponse {
  final bool restarted;
  final int segmentFrom;
  final int? actualStartMs;
  SeekResponse({required this.restarted, required this.segmentFrom, required this.actualStartMs});
  factory SeekResponse.fromJson(Map<String, dynamic> j) => SeekResponse(
        restarted: j['restarted'] as bool,
        segmentFrom: j['segmentFrom'] as int,
        actualStartMs: j['actualStartMs'] as int?,
      );
}

class AudioTrackSwitchResponse {
  final bool restarted;
  final int segmentFrom;
  final int? actualStartMs;
  AudioTrackSwitchResponse({required this.restarted, required this.segmentFrom, required this.actualStartMs});
  factory AudioTrackSwitchResponse.fromJson(Map<String, dynamic> j) => AudioTrackSwitchResponse(
        restarted: j['restarted'] as bool,
        segmentFrom: j['segmentFrom'] as int,
        actualStartMs: j['actualStartMs'] as int?,
      );
}

class MediaItemRef {
  final String id;
  final String kind;
  final String title;
  final String? parentId;
  final int? seasonNumber;
  final int? episodeNumber;
  final int? year;
  final String? posterUrl;
  final String? backdropUrl;
  final String? mediaFileId;

  MediaItemRef({
    required this.id,
    required this.kind,
    required this.title,
    required this.parentId,
    required this.seasonNumber,
    required this.episodeNumber,
    required this.year,
    required this.posterUrl,
    required this.backdropUrl,
    required this.mediaFileId,
  });

  factory MediaItemRef.fromJson(Map<String, dynamic> j) => MediaItemRef(
        id: j['id'] as String,
        kind: j['kind'] as String,
        title: j['title'] as String,
        parentId: j['parentId'] as String?,
        seasonNumber: j['seasonNumber'] as int?,
        episodeNumber: j['episodeNumber'] as int?,
        year: j['year'] as int?,
        posterUrl: j['posterUrl'] as String?,
        backdropUrl: j['backdropUrl'] as String?,
        mediaFileId: j['mediaFileId'] as String?,
      );
}

class ContinueWatchingEntry {
  final MediaItemRef mediaItem;
  final String? seriesTitle;
  final String detailItemId;
  final int positionMs;
  final int? durationMs;
  final bool upNext;

  ContinueWatchingEntry({
    required this.mediaItem,
    required this.seriesTitle,
    required this.detailItemId,
    required this.positionMs,
    required this.durationMs,
    required this.upNext,
  });

  factory ContinueWatchingEntry.fromJson(Map<String, dynamic> j) => ContinueWatchingEntry(
        mediaItem: MediaItemRef.fromJson(j['mediaItem'] as Map<String, dynamic>),
        seriesTitle: j['seriesTitle'] as String?,
        detailItemId: j['detailItemId'] as String,
        positionMs: j['positionMs'] as int,
        durationMs: j['durationMs'] as int?,
        upNext: j['upNext'] as bool? ?? false,
      );
}
